libbpf-rs/src/program.rs

```diff
@@ -864,6 +930,136 @@ impl<'obj> ProgramMut<'obj> {
+                &opts as *const _,
+            )
+        };
+        let ptr = validate_bpf_ret(ptr).context("failed to attach kprobe multi")?;
+        // SAFETY: the pointer came from libbpf and has been checked for errors.
+        let link = unsafe { Link::new(ptr) };
+        Ok(link)
+    }
+
+    /// Attach this program to multiple [kernel
+    /// probes](https://www.kernel.org/doc/html/latest/trace/kprobetrace.html)
+    /// at once.
+    pub fn attach_kprobe_multi<T: AsRef<str>>(
+        &self,
+        symbols: Vec<T>,
+        retprobe: bool,
+    ) -> Result<Link> {
+        let cnt = Self::check_kprobe_multi_args(&symbols, &[])?;
+
+        let mut syms = symbols
+            .iter()
+            .map(|s| util::str_to_cstring(s.as_ref()).map(|s| s.into_raw()))
+            .collect::<Result<Vec<_>>>()?;
+
+        let opts = libbpf_sys::bpf_kprobe_multi_opts {
+            sz: size_of::<libbpf_sys::bpf_kprobe_multi_opts>() as _,
+            syms: syms.as_mut_ptr() as _,
+            cnt: cnt as libbpf_sys::size_t,
+            retprobe,
+            // bpf_kprobe_multi_opts might have padding fields on some platform
+            ..Default::default()
+        };
+
+        let link = self.attach_kprobe_multi_impl(opts)?;
+
+        // Calling `CString::into_raw` requires the CString to be reconstituted
+        // using `CString::from_raw` to be properly deallocated.
+        let _ = syms.drain(..).map(|s| unsafe { CString::from_raw(s) });
+
+        Ok(link)
+    }
+
+    /// Attach this program to multiple [kernel
+    /// probes](https://www.kernel.org/doc/html/latest/trace/kprobetrace.html)
+    /// at once, providing additional options.
+    pub fn attach_kprobe_multi_with_opts(&self, opts: KprobeMultiOpts) -> Result<Link> {
+        let KprobeMultiOpts {
+            symbols,
+            mut cookies,
+            retprobe,
+            _non_exhaustive,
+        } = opts;
+
+        let cnt = Self::check_kprobe_multi_args(&symbols, &cookies)?;
+
+        let mut syms = symbols
+            .iter()
+            .map(|s| util::str_to_cstring(s.as_ref()).map(|s| s.into_raw()))
+            .collect::<Result<Vec<_>>>()?;
+
+        let opts = libbpf_sys::bpf_kprobe_multi_opts {
+            sz: size_of::<libbpf_sys::bpf_kprobe_multi_opts>() as _,
+            syms: syms.as_mut_ptr() as _,
+            cookies: if !cookies.is_empty() {
+                cookies.as_mut_ptr() as _
+            } else {
+                ptr::null()
+            },
+            cnt: cnt as libbpf_sys::size_t,
+            retprobe,
+            // bpf_kprobe_multi_opts might have padding fields on some platform
+            ..Default::default()
+        };
+
+        let link = self.attach_kprobe_multi_impl(opts)?;
+
+        // Calling `CString::into_raw` requires the CString to be reconstituted
+        // using `CString::from_raw` to be properly deallocated.
+        let _ = syms.drain(..).map(|s| unsafe { CString::from_raw(s) });
```

Comment:
I am confused about this line.

Reply:
`CString::into_raw` transfers ownership and to avoid memory leak the above is required. This is explained in the doc, https://doc.rust-lang.org/std/ffi/struct.CString.html#method.into_raw.

Reply:
It makes sense, but it's kind of error prone. The next person may adjust the `attach` call above to return eagerly and we end up leaking again (even more easy to do because the `link` is actually a `Result<Link>` and not `Link` as seems to be implied). Personally I'd prefer we keep the `Vec<CString>` around and then another `Vec` comprised of a mapping using the non-consuming https://doc.rust-lang.org/std/ffi/struct.CString.html#method.as_ptr

I don't feel strongly, though.

Reply:
Actually, we may leak even now: if part of
```rust
let mut syms = symbols
            .iter()
            .map(|s| util::str_to_cstring(s.as_ref()).map(|s| s.into_raw()))
```
fails midway through then I think we loose the first part that has already been converted to raw pointers, right? So we should probably fix that up.

Reply:
Those are very good points; you're right choosing `into_raw` instead of `as_ptr` to avoid keeping a `Vec<CString>` around led to leak memory on error. Changing this as you suggested, keeping an extra `Vec` around.